File: web/tests/reference/queues/stream-first-unique.js
class FirstUnique {
  constructor() {
    this.count = new Map();
    this.q = [];
    this.head = 0;
  }

  add(ch) {
    this.count.set(ch, (this.count.get(ch) || 0) + 1);
    this.q.push(ch);
    while (this.head < this.q.length && this.count.get(this.q[this.head]) > 1) {
      this.head++;
    }
    return this.head < this.q.length ? this.q[this.head] : "#";
  }
}

File: web/tests/reference/queues/stream-first-unique.py
from collections import deque

class FirstUnique:
    def __init__(self):
        self.count = {}
        self.q = deque()

    def add(self, ch):
        self.count[ch] = self.count.get(ch, 0) + 1
        self.q.append(ch)
        while self.q and self.count[self.q[0]] > 1:
            self.q.popleft()
        return self.q[0] if self.q else "#"
